File: python/dsa/hashing/bloom_filter.py
# Hashing — Bloom Filter

"""Probabilistic set membership: no false negatives, possible false positives.

add/might_contain O(k) where k is the number of hash functions.
"""

from __future__ import annotations

import math


class BloomFilter:
    """Bloom filter for string keys."""

    def __init__(self, expected_items: int, false_positive_rate: float) -> None:
        """Creates a bloom filter with capacity and false-positive rate target.

        Time: O(m); Space: O(m).
        """
        self._size = _optimal_size(expected_items, false_positive_rate)
        self._num_hashes = _optimal_hashes(self._size, expected_items)
        self._bits = [False] * self._size

    def add(self, item: str) -> None:
        """Adds an element (string key).

        Time: O(k); Space: O(1).
        """
        for i in range(self._num_hashes):
            self._bits[self._hash(item, i)] = True

    def might_contain(self, item: str) -> bool:
        """Returns True if the item might be present (false positives possible).

        Returns False if definitely absent.
        Time: O(k); Space: O(1).
        """
        return all(self._bits[self._hash(item, i)] for i in range(self._num_hashes))

    def _hash(self, item: str, seed: int) -> int:
        h = hash(item) ^ (seed * 0x9E3779B9)
        return h % self._size


def _optimal_size(n: int, p: float) -> int:
    return max(8, math.ceil(-n * math.log(p) / (math.log(2) ** 2)))


def _optimal_hashes(m: int, n: int) -> int:
    return max(1, round(m / n * math.log(2)))

File: java/src/main/java/com/dsa/examples/hashing/BloomFilter.java
// Hashing — Bloom Filter

package com.dsa.examples.hashing;

/**
 * Probabilistic set membership: no false negatives, possible false positives.
 * add/contains O(k) where k is the number of hash functions.
 */
public class BloomFilter {

    private final boolean[] bits;
    private final int numHashes;
    private final int size;

    /**
     * Creates a bloom filter with {@code expectedItems} capacity and false-positive rate target.
     * Time: O(m); Space: O(m).
     */
    public BloomFilter(int expectedItems, double falsePositiveRate) {
        this.size = optimalSize(expectedItems, falsePositiveRate);
        this.numHashes = optimalHashes(size, expectedItems);
        this.bits = new boolean[size];
    }

    /**
     * Adds an element (string key).
     * Time: O(k); Space: O(1).
     */
    public void add(String item) {
        for (int i = 0; i < numHashes; i++) {
            bits[hash(item, i)] = true;
        }
    }

    /**
     * Returns true if the item might be present (false positives possible).
     * Returns false if definitely absent.
     * Time: O(k); Space: O(1).
     */
    public boolean mightContain(String item) {
        for (int i = 0; i < numHashes; i++) {
            if (!bits[hash(item, i)]) {
                return false;
            }
        }
        return true;
    }

    private int hash(String item, int seed) {
        int h = item.hashCode() ^ (seed * 0x9e3779b9);
        return Math.floorMod(h, size);
    }

    private static int optimalSize(int n, double p) {
        return Math.max(8, (int) Math.ceil(-n * Math.log(p) / (Math.log(2) * Math.log(2))));
    }

    private static int optimalHashes(int m, int n) {
        return Math.max(1, (int) Math.round((double) m / n * Math.log(2)));
    }
}
